### PR TITLE
feat: add read-only transactions

### LINKED ISSUE #1245: Make it possible to use read-only transactions
The runTransaction call creates a read-write transaction that acquires locks on the data. This is a problem when the transaction is both long-lived (so would hold locks for a long time) and reads a lot of data (or data which is frequently written), as it will delay concurrent writes.

The Firestore API offers read-only transactions as an alternative. These do not acquire locks and do not impact writes.

A plausible way to add these is to add a 'read-only' option to runTransaction, or to add a runReadOnlyTransaction call.

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds read-only transactions to Firestore Node. They are already available in Java.

Fixes https://github.com/googleapis/nodejs-firestore/issues/1245

### CODE REVIEW

File: types/firestore.d.ts [hunk 1]
```diff
@@ -311,25 +333,30 @@ declare namespace FirebaseFirestore {
...
+     * Transaction can be performed as read-only or read-write transactions. By
+     * default, transactions are executed in read-write mode.
+     *
+     * A read-write transaction obtains a pessimistic lock on all documents that
+     * are read during the transaction. These locks block other transactions,
+     * batched writes, and other non-transactional writes from changing that
+     * document. A transaction releases its document locks at commit time or if
+     * it fails for any reason.
+     *
+     * Read-only transaction do not lock documents. They can be used to read
```
Comment (thebrianchen) [inline types/firestore.d.ts:345]: s/transaction/transactions

File: dev/src/index.ts [hunk 2]
```diff
@@ -945,26 +970,33 @@ export class Firestore implements firestore.Firestore {
    * modify Firestore documents under lock. You have to perform all reads before
    * before you perform any write.
    *
-   * Documents read during a transaction are locked pessimistically. A
-   * transaction's lock on a document blocks other transactions, batched
-   * writes, and other non-transactional writes from changing that document.
-   * A transaction releases its document locks at commit time or once it times
-   * out or fails for any reason.
+   * Transaction can be performed as read-only or read-write transactions. By
```
Comment (thebrianchen) [inline dev/src/index.ts:973]: s/Transaction/Transactions

File: dev/src/index.ts [hunk 3]
```diff
@@ -995,28 +1027,54 @@ export class Firestore implements firestore.Firestore {
...
+      if (transactionOptions.readOnly) {
+        validateTimestamp(
+          'transactionOptions.readTime',
+          transactionOptions.readTime,
+          {optional: true}
+        );
+
+        readOnly = true;
+        readTime = transactionOptions.readTime as Timestamp | undefined;
+        maxAttempts = 1;
```
Comment (thebrianchen) [inline dev/src/index.ts:1059]: optional: specify that read-only transactions not retried in the documentation.

File: dev/system-test/firestore.ts [hunk 4]
```diff
@@ -2345,6 +2345,41 @@ describe('Transaction class', () => {
...
+    await ref.set({foo: 2});
+    const snapshot = await firestore.runTransaction(
+      updateFunction => updateFunction.get(ref),
+      {readOnly: true, readTime: writeResult.writeTime}
+    );
+    expect(snapshot.exists).to.be.true;
+    expect(snapshot.get('foo')).to.equal(1);
+  });
+
+  it('fails read-only with writes', async () => {
```
Comment (thebrianchen) [inline dev/system-test/firestore.ts:2371]: optional: verify that the tx was not retried.

File: types/firestore.d.ts [hunk 5]
```diff
@@ -174,6 +174,28 @@ declare namespace FirebaseFirestore {
     [key: string]: any; // Accept other properties, such as GRPC settings.
   }
 
+  /** Options to configure a read-only transaction. */
+  export interface ReadOnlyTransactionOptions {
+    /** Set to true to indicate a read-write transaction. */
```
Comment (thebrianchen) [inline types/firestore.d.ts:179]: s/read-write/read-only

File: types/firestore.d.ts [hunk 6]
```diff
@@ -174,6 +174,28 @@ declare namespace FirebaseFirestore {
...
+     * If specified, documents are read at the given time. This may not be more
+     * than 60 seconds in the past from when the request is processed by the
+     * server.
+     */
+    readTime?: Timestamp;
+  }
+
+  /** Options to configure a read-write transaction. */
+  export interface ReadWriteTransactionOptions {
+    /** Set to false or omitted to start a read-write transaction. */
```
Comment (thebrianchen) [inline types/firestore.d.ts:191]: s/start/indicate

**Review by thebrianchen — APPROVED** [review]

File: dev/src/index.ts (hunk 3 above)
Comment (schmidt-sebastian) [reply dev/src/index.ts:1058]: Done (might make sense for you to do a quick sanity check on the result)

File: dev/system-test/firestore.ts (hunk 4 above)
Comment (schmidt-sebastian) [reply dev/system-test/firestore.ts:2371]: "Sebastian goes above and beyond"

File: types/firestore.d.ts (hunk 5 above)
Comment (schmidt-sebastian) [reply types/firestore.d.ts]: Done

File: types/firestore.d.ts (hunk 6 above)
Comment (schmidt-sebastian) [reply types/firestore.d.ts]: done

File: types/firestore.d.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply types/firestore.d.ts]: Done

**Pull request merged**